crates/polars-core/src/series/any_value.rs

```diff
@@ -449,103 +451,60 @@ impl Series {
-            None => {
-                if all_flat_null {
-                    Ok(Series::new_null(name, avs.len()))
-                } else {
-                    // second pass and check for the nested null value that toggled `all_flat_null` to false
-                    // e.g. a list<null>
-                    if let Some(av) = avs.iter().find(|av| !matches!(av, AnyValue::Null)) {
-                        let dtype: DataType = av.into();
-                        Series::from_any_values_and_dtype(name, avs, &dtype, strict)
+    /// Construct a new [`Series`] from a slice of AnyValues.
+    ///
+    /// The data type of the resulting Series is determined by the `values`
+    /// and the `strict` parameter:
+    /// - If `strict` is `true`, the data type is equal to the data type of the
+    ///   first non-null value. If any other non-null values do not match this
+    ///   data type, an error is raised.
+    /// - If `strict` is `false`, the data type is the supertype of the
+    ///   `values`. The supertype is determined by a pass over the values.
+    ///   Values encountered that do not match the supertype are set to null.
+    /// - If no values were passed, the resulting data type is `Null`.
+    pub fn from_any_values(name: &str, values: &[AnyValue], strict: bool) -> PolarsResult<Series> {
+        fn get_first_non_null_dtype(values: &[AnyValue]) -> DataType {
+            let mut all_flat_null = true;
+            let first_non_null = values.iter().find(|av| {
+                if !av.is_null() {
+                    all_flat_null = false
+                };
+                !av.is_nested_null()
+            });
+            match first_non_null {
+                Some(av) => av.dtype(),
+                None => {
+                    if all_flat_null {
+                        DataType::Null
                     } else {
-                        unreachable!()
+                        // Second pass to check for the nested null value that
+                        // toggled `all_flat_null` to false, e.g. a List(Null)
+                        let first_nested_null = values.iter().find(|av| !av.is_null()).unwrap();
+                        first_nested_null.dtype()
                     }
-                }
-            },
-            Some(av) => {
-                #[cfg(feature = "dtype-decimal")]
-                {
-                    if let AnyValue::Decimal(_, _) = av {
-                        let mut s = any_values_to_decimal(avs, None, None)?.into_series();
-                        s.rename(name);
-                        return Ok(s);
+                },
+            }
+        }
+        fn get_any_values_supertype(values: &[AnyValue]) -> DataType {
+            let mut supertype = DataType::Null;
+            let mut dtypes = PlHashSet::<DataType>::new();
+            for av in values {
+                if dtypes.insert(av.dtype()) {
+                    supertype = match try_get_supertype(&supertype, &av.dtype()) {
```

Comment:
instead of `try_get_supertype` we should use `get_supertype` directly here. It returns an `Option` meaning we don't allocate a string on the error case.